### PR TITLE
Improve and refactor derive logic.

### PR DESCRIPTION
**Pull request by DelSkayn**
Overhauls the derive macro to better support removing fields in a enum variant and removing the fields of types which are no longer used in the latest revision.

This PR introduces field structs for enum variants.
When a field is removed from an enum variant, the convert function now looks like `fn(&mut FieldStruct,revision: u16, value: T) -> Result<()>` where `FieldStruct` is a generated struct which has the same fields as the variant has at the current revision.
Furthermore when a variant is removed the convert function now takes this field struct as an argument.

This PR also introduces an option to override the default generated discriminant values for enum variants. This is useful for compatibility with previous versions of revision which had different behavior.

### CODE REVIEW

File: tests/test.rs [hunk 1]
```diff
@@ -19,64 +19,68 @@ pub enum TestEnum {
 		a: i64,
 		#[revision(end = 3, convert_fn = "upgrade_three_b")]
 		b: f32,
-		#[revision(start = 3)]
+		#[revision(start = 3, default_fn = "default_three_c")]
 		c: f64,
-		#[revision(start = 3)]
+		#[revision(start = 3, default_fn = "default_three_d")]
 		d: String,
```
Comment (emmanuel-keller) [inline tests/test.rs:25]: If supported, shall we have a test where, in a variant, we change a field from a tuple?

```Rust
ThreeTuple (
		#[revision(end = 3, convert_fn = "upgrade_three_1")]
		i32,
		#[revision(start = 3, default_fn = "default_three_2")]
		i64
)
```

Comment (DelSkayn) [reply tests/test.rs:25]: If you see the whole file you will notice there already is a test which does that:
```rust
#[revisioned(revision = 2)]
#[derive(Debug, Default, PartialEq)]
pub struct TestTuple2(
	#[revision(end = 2, convert_fn = "convert_tuple")] pub Vec<i64>,
	#[revision(start = 2)] pub Vec<f64>,
);
```

Comment (DelSkayn) [reply tests/test.rs:25]: That is supported I will add the test for it.

Comment (DelSkayn) [reply tests/test.rs:25]: That also immediately revealed a bug so good suggestion!

**Review by emmanuel-keller — APPROVED** [review]

**Review by rushmorem — APPROVED** [review]

**Review by kearfy — APPROVED** [review]

**Review by tobiemh — APPROVED** [review]

**Pull request merged**